### ISSUE TITLE
Why is unlimited storage permissions for update required?

### ISSUE THREAD

**Issue by the-j0k3r** [EFForg/privacybadger#1745]
Ive just received a message in Firefox that to update to the new Privacy badger, it now requires unlimited storage permissions.

A cursive search reveals this was added at https://github.com/EFForg/privacybadger/pull/1742

WHY unlimited storage? I dont want anything to have such permission, its just too broad and ripe for abuse.

Is there no way around this like making it optional for instance?
Is it possible to allow a flexible storage option that user can configure?

**Comment by Videonauth** [comment]
Same question from me, why does a tool which should prevent ad sites and trackers getting data of me now want to store data itself. This is a paradox pure and simple. Do I replace now many evil ad's and trackers with another evil?

**Comment by ghostwords** [comment]
Privacy Badger requires storing which domains track you (and where the tracking took place) in order to function. Extension storage ([`chrome.storage.local`](https://developer.chrome.com/apps/storage)) is by default limited to ~5 MB. A user ran over this limit (#1717); adding `unlimitedStorage` was the most straightforward way to deal with the immediate problem.

I understand you are concerned about Privacy Badger using too much space. You can check how much space (approximately) your Badger uses now by running the following code in your Badger's background page:

```javascript
chrome.storage.local.get(null, r=>console.log(
  parseFloat(JSON.stringify(r).length / 1024 / 1024).toFixed(2), "MB"));
```

To get to the background page console in Firefox, visit `about:debugging`, enable add-on debugging, click Debug next to Privacy Badger, click the OK button on the popup warning about remote debugging, and enter the above script into the console after the >>.

We plan on reducing and optimizing what Badger stores in the future (most likely starting with #1446).

**Comment by Videonauth** [comment]
So this is about local storage only? I'm not so concerned if I store my own data on my own drive including browsing and do not track settings. I was more concerned if that data is getting transmitted as well. And to be honest seeing something pop-up in firefox saying something wants the permission to store an unlimited amount of client-data, simply gets my alarm bells ringing. (see attached screen-shot)
![selection_007](https://user-images.githubusercontent.com/9640109/32060393-e5b763a6-ba6e-11e7-8dfc-1eebe8277dfa.png)

**Comment by ghostwords** [comment]
Strangely, we used to have the "unlimitedStorage" permission, but removed it in 67588e00a518eda2fcface97ff1b3a3723bf2c62. We might have overreacted to warnings from Firefox dev tools (https://github.com/EFForg/privacybadgerfirefox-legacy/issues/208#issuecomment-239054130).

uBlock Origin declares "unlimitedStorage" as well (and documents Privacy Badger as having it): https://github.com/gorhill/uBlock/wiki/About-the-required-permissions.

Adding the "unlimitedStorage" permission does not produce any additional warnings in Chrome (https://github.com/EFForg/privacybadger/issues/1717#issuecomment-338282900). You could argue this is another breaking change in Firefox's implementation of WebExtensions.

**Comment by ghostwords** [comment]
@Videonauth Correct, this data stays on your local machine. Agreed, Firefox's notification provokes needless concern. See my above comment regarding how this notification shouldn't even exist in Firefox.

**Comment by Videonauth** [comment]
Yes thank you this makes it more clear, in fact this pop-up made me come here, I honestly wonder why this thread not already has thousands of people in it. My first thought was hell what who is go and store that unlimited amount of data and foremost what data is going to be stored :) I changed from ABP to badger because I want no one to have my data. Your last text for sure made it clear, Thank you.

**Comment by ghostwords** [comment]
I opened a Bugzilla issue for Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1411999.

**Comment by jawz101** [comment]
Yeah.  It's really an awkward prompt for a user for that particular thingy.  Probably doesn't evoke what would be a significant change to an add-on.  On the other hand, doing so may give pause for add-on devs to consider tinier ways to achieve the same purpose in an add-on without relying on large data sets.

**Comment by SuperSluether** [comment]
Does uBlock Origin really use unlimited storage? It's not listed in the permissions on the add-on page: https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/

I really don't think this is a valid Bugzilla issue, and I agree with Andy's reply. If an Android app updates to include new permissions, the user is notified before installing the update. I believe Firefox should continue displaying prompts like this to let users know what is happening.

The prompt is clear that it is _client-side_ storage. My browsing history is stored client-side, so a message like that shouldn't raise any red flags.

WebExtensions may be based on Chrome, but I'm glad Firefox is doing some things differently.

**Comment by jawz101** [comment]
woah.  How did you get a real name from a username on github? ... or maybe it's someone else 0_o

**Comment by SuperSluether** [comment]
@jawz101 Andy McKay from the Bugzilla link: https://bugzilla.mozilla.org/show_bug.cgi?id=1411999

**Comment by jawz101** [comment]
ooh... well my name is actually Zotan The Destroyer.  Hope that clears things up :D

**Comment by ghostwords** [comment]
@SuperSluether Yeah, I'm not sure what's going on with uBlock Origin on Firefox, but uBlock Origin on Chrome does declare "unlimitedStorage".

**Comment by the-j0k3r** [comment]
Im still unhappy that unlimited storage may be abused and would prefer a more hands on user side control on this.

> You could argue this is another breaking change in Firefox's implementation of WebExtensions.

I could argue a great deal on how Mozilla is killing Firefox by removing what made it stand out in first place, I could argue but I wont.

please consider allowing user to set own limits rather than  current design, it would be nice if you just get notified, "hey I need a few more MB to store your data".

**Comment by ghostwords** [comment]
@the-j0k3r How much storage does your Badger use now (https://github.com/EFForg/privacybadger/issues/1745#issuecomment-339691221)?

**Comment by SuperSluether** [comment]
"I could argue a great deal on how Mozilla is killing Firefox by removing what made it stand out in first place, I could argue but I wont."

Then don't. Because they aren't. Add-ons are in a much better place now, people need to get over themselves.

Back to the topic, are you worried about unlimited storage being abused by PB, or by add-ons in general?

**Comment by the-j0k3r** [comment]
@SuperSluether thanks for your interest, Ive put my concerns already forward, there is nothing to be added of value by repeating myself.

**Comment by SuperSluether** [comment]
@the-j0k3r When you said

> its just too broad and ripe for abuse.

I'm wondering what exactly you mean. It seems to me that Privacy Badger would already be ripe for abuse, as it can access your browsing data and block connections. How exactly would PB abuse storing its own options?

No need to repeat "ripe for abuse" or "may be abused". I'm just wondering what "abuse" it is you are worried about.

**Comment by the-j0k3r** [comment]
> @the-j0k3r How much storage does your Badger use now (#1745 (comment))?

0.20 MB

**Comment by stemid** [comment]
I'm sorry but you made a boo boo here. You opted to allow a firefox extension unlimited storage instead of advising users with this issue to change dom.storage.default_quota in their browser to a higher value. 

That's opening the gates for future exploitation imho. 

A minority of users are filling their local storage so everyone must now allow an extension unlimited storage to solve that problem. 

We can't say how this would be exploited but one guess is to simply randomize domains loaded until the storage is filled, like a DoS against the client. I have no insight into the code base but I feel that this issue is more about common sense.

Point is that unlimited storage is unnecessarily dangerous.

**Comment by gorhill** [comment]
> Does uBlock Origin really use unlimited storage?

No, I removed the permission before releasing uBO/webext because it was causing a warning of unsupported permission at the browser console back then. Apparently support for `unlimitedStorage` was added recently. uBO uses indexedDB to store its large chunks of data (which sum up to beyond 5 MB) -- I don't know if declaring the `unlimitedStorage` permission will be required by uBO for using indexedDB.

To those who freak out about this unlimited storage permission, keep in mind that with _all_ legacy extensions there was no limit whatsoever. The only difference is that now you are _informed_ about extensions which requires unlimited storage (you were not before). So to see this current permission as being an issue makes no sense considering how things were before WebExtensions.

**Comment by the-j0k3r** [comment]
💯 @stemid is on the right track, especially on a day and age where you read multiple news daily about lesser holes being exploited by nefarious and unscrupulous 3rd parties in exotic ways to achieve their goals.

@gorhill it doesn't matter what the previous state of affairs was in reality, I agree that the previous situation was not right but I wouldn't use that as an argument to dismiss the issue, which is pretty much how that sounds.

**Comment by SuperSluether** [comment]
@the-j0k3r I believe what @gorhill is getting at is that this is nothing new, and not anything to be concerned about. If you don't want Privacy Badger to "exploit" storing it's own data in some unforeseen way that you still haven't explained, then don't use Privacy Badger.

Hell, don't use a browser either. Those things get exploited all the time. Should probably stop using computers too.

**Comment by SuperSluether** [comment]
I'm getting bored of this circular discussion. The explanation wasn't needed in the first place, as the original question was answered by #1717. 

Both Firefox and Privacy Badger are working as intended. Privacy Badger needs to store its options, Firefox lets you know it's storing data.

Working as intended, won't fix, whatever. If you decide to "fix" it, just make sure it doesn't break for all us non-paranoids, Kay?

Thanks bye.

**Comment by ghostwords** [comment]
Just a reminder, while it's fine to disagree, let's please remain respectful and constructive as we do so.

**Comment by ghostwords** [comment]
Let's wait for the 2017.10.25.1 update to finish rolling out, and then replace this issue with a new issue to remove the "unlimitedStorage" permission once Badger becomes unlikely to exceed 5 MB through regular use (after #1446 and related issues).

As https://github.com/EFForg/privacybadger/issues/1745#issuecomment-339914562 points out, "unlimitedStorage" probably does introduce a risk of somebody messing with your browser by making Badger learn a lot of garbage data. We should look into it anyway.

**Comment by the-j0k3r** [comment]
@ghostwords #1747 is not an impatience driven issue, Just posted now is all.

This topic is a good place to discuss the semantics, will look forward to future mitigation improvements. 

Thank you,

**Comment by gorhill** [comment]
@the-j0k3r 

> it doesn't matter what the previous state of affairs was, in reality I agree that the previous situation was not right

I use the argument to point out that blaming Privacy Badger _now_ makes no sense, things were worst before when you were not informed. If you feel the "unlimited storage" permission is wrong, you need to report to Bugzilla, this is a WebExtensions feature, which extensions use at their discretion.

**Comment by stemid** [comment]
@gorhill and @SuperSleuther Is there a reason to not just use dom.storage.default_quota in firefox?

**Comment by SuperSluether** [comment]
@stemid I prefer installing an add-on once and having it work as intended. I don't want to have to change things in about:config when the add-on can just use the unlimited storage permission.

I especially don't want to have to explain about:config to my non-tech-savy parents/friends if they happen to hit the limit. It's much easier to show them how to click "Add to Firefox" and not worry about something breaking.

**Comment by stemid** [comment]
Understandable but in this case I think it would be easier to make that config change once a user hits that limit than blanket all users with what Firefox devs and others seem to consider an invasive permission. 

Even if we're only just now seeing these warnings it's still relevant to a group of privacy conscious users, just like android permissions and other things in that vein.

**Comment by the-j0k3r** [comment]
Why would you need to change addon specific settings in about:config which is a browser configuration thing? You dont/wouldnt simpy put.

@stemid agreed.

@SuperSluether its abundantly clear you disapprove and disagree with the issue of this topic. Just because your friends are less technically minded is again an excuse. People should be educated to understand that using 12345 as password is wrong and that allwoing things that can be abused in reality is also wrong,

> just because you can buy a hammer to hammer down a nail or five, it doenst mean its right to use that hammer to crack someone's head, yet things are misused all the time.

I refer back to my earlier hammer analogy to extend that if you saw your friends hammering their own fingers, I would like to believe you would teach them how to use the hammer, or if you saw someone trying to crack your friends head with that hammer, again Ide like to believe you would try to stop them.

Yes Im comparing this unlimited storage to a hammer, because it acts like a hammer to resolve a extreme use-case issue until better garbage collection will make it possible to disable in future.

**Comment by SuperSluether** [comment]
> allwoing things that can be abused in reality is also wrong

I guess I should start telling everyone to uninstall their browsers and disconnect their Internet modems, because by using the Internet on a smartphone or computer they are allowing things that can be abused.

By extension, posting this message is also wrong, since I'm using a computer, browser, internet connection, and web platform that can all be abused.

I guess I'm just a moron for thinking the Internet could be used for anything other than bad. How dare I think that storing data is a reasonable thing for a browser add-on to do.

**Comment by the-j0k3r** [comment]
@gorhill for implementing similar in ublock origin
